### DISCUSSION TITLE
Goal Setting indefinitely.  Unable to autonomously navigate

### DISCUSSION THREAD

**Discussion by timothynassar** [prl-mushr/mushr#75] [Software]
I currently have a mushr robot built to the specifications of the mushr website.  I am working on setting goals in rviz with a map that I have made.  When I set the goal in rviz, the console notifies me that the goal is setting, but it never notifies me that the goal is set.
![image](https://user-images.githubusercontent.com/86331107/158871165-bee7854c-8173-44b6-8d3c-3850e10b002c.png)  I have tried adjusting the value for K in dispersion.yaml to 30 and I have changed the sampling size in the RHC controller to 11 as recommended in a previous discussion on this forum.  I would appreciate any advice.  Thank you

**Comment by schmittlema** [comment]
Unfortunately mushr_rhc is in a broken state and we are working to re-write it to into more readible and usable form. My best recommendation in the meantime is to try following the instructions layed out [here](https://github.com/prl-mushr/mushr/discussions/70#discussioncomment-2150556).

**Reply by timothynassar** [reply]
Thank you for the help.  We are having some issues with editing the controlnode.py file.  We have edited lines 84-95 in the following picture, but we are still getting errors.  Do you see anything wrong with the modifications we have made? 

![image](https://user-images.githubusercontent.com/86331107/159064770-67476757-c0b0-4aa2-8d11-334c239af233.png)

**Reply by schmittlema** [reply]
What error are you getting?

**Reply by timothynassar** [reply]
We are getting this error after attempting to set a 2D Nav Goal in Rviz

![image](https://user-images.githubusercontent.com/86331107/159073979-a612d962-6eaa-41f4-ab33-64d2453d3b79.png)

**Reply by schmittlema** [reply]
so your cb_path function is expecting a path not a single point (PoseStamped). You will want to change that function to set `path = msg.pose` and that will likely cause some further errors downstream.

**Reply by timothynassar** [reply]
We are getting an indexing error now when the program starts indexing the for loop in visualize_path? Any advice from this point?  We got to a similar point before with some of our own troubleshooting, but couldn't figure out how to repair the issue from this point forward.  
![image](https://user-images.githubusercontent.com/86331107/159099888-4c2468b0-916b-42e5-83bc-1945e76aeb00.png)

**Reply by schmittlema** [reply]
So from the looks of this error line 195 is trying to index the path as if it were an array. It may be easiest to change `path = msg.pose` from the previous message to `path = [msg.pose]` to make it an array object that way it is treated as a single point "path"

**Reply by timothynassar** [reply]
After making that change, we got this error.  Could this possible have to do with an incorrect marker topic selection in rviz or is this related to the python script?

![image](https://user-images.githubusercontent.com/86331107/159364623-b16693b0-d88a-4340-9f80-f2a77c50451f.png)

**Reply by schmittlema** [reply]
These errors are all related to adapting the python script to work with a single goal point. The [PoseStamped object](http://docs.ros.org/en/noetic/api/geometry_msgs/html/msg/PoseStamped.html) you are subscribed to contains a pose object which holds a position/orientation. The script is expecting a different type of object but with proper adaptation it will accept a pose object. Try replacing the line I mentioned to the following. You may get an error about orientation but let's see:

`path = [msg.pose.position]`

**Reply by timothynassar** [reply]
Ok yes, we did get an error related to the XYHV position.  What might be the next step from here? 

![image](https://user-images.githubusercontent.com/86331107/159510368-a721834e-f542-4356-a514-f627a65f97a9.png)

**Reply by schmittlema** [reply]
So "h" is the robot's heading and "v" is the desired velocity. In this case I would recommend recommend replacing `path[i].h` with 0 and `path[i].v` to 1.5. These hard coded values will at least let you continue and hopefully get something running.  Afterward we will want to go back and convert `pose.orientation` which is a quaternion into a euler heading (yaw) to get a proper desired heading.

**Reply by timothynassar** [reply]
Thank you so much! We were finally able to get the path to set.  The only problem is that once the path is set, we hit the R1 dead-man switch on the car, but there is no movement.  Is there a different sequence to get the robot moving with ta_lab2-master than the dead-man switch? Could it be that the v parameter of 1.5 is too low?

**Reply by schmittlema** [reply]
Are you running this in sim or on the real robot? 
If you are running in sim:
Line 89 in `controlnode.py` needs to subscribe to the right position topic. `/sim_car_pose` needs to be `/car/car_pose`

If you are running on the real robot:
Are you running the particle filter `mushr_pf`?
If so Line 89 needs to be `/car/particle_filter/inferred_pose`

**Reply by timothynassar** [reply]
We are attempting to run on the real robot.  We set line 89 as you said, but got an error related to this new topic selection.

![image](https://user-images.githubusercontent.com/86331107/159968757-2d67feb0-534e-4acc-a276-ec7e442db4f4.png)

![image](https://user-images.githubusercontent.com/86331107/159969453-993a350c-9d23-4697-9b6e-155eccb8a141.png)

**Reply by schmittlema** [reply]
If you remove the `rospy.get_param()` and keep the string inside it then I think it should work. If you want to learn more about ROS I encourage you to try the[ following tutorial](http://wiki.ros.org/ROS/Tutorials/WritingPublisherSubscriber%28python%29) to get yourself familiar.

**Reply by timothynassar** [reply]
Ok we made that change and got the following error about ip around line 53 of controlnode.py and you can see how that error was remedied in the picture below.

![image](https://user-images.githubusercontent.com/86331107/160181399-93f5b44f-299f-406e-92ca-45dea3833d72.png)

Once that issue was mitigated, we were able to get the path and set the path, but still no movement upon engagement of the dead-man switch.  In the photo below you can see that we are running without vital errors, but we get no movement :/

![image](https://user-images.githubusercontent.com/86331107/160181878-e3b728f2-39b3-4f8a-97c8-3b88ccbe8291.png)

**Reply by bears0** [reply]
Is there a working version of the RHC node? I imagine it worked before, so how would we go about getting a previous stable release?

**Reply by timothynassar** [reply]
We were actually able to finally get the waypoint navigation to work and my group will share the revision that works.  The problem we are still having is that the robot doesn’t avoid obstacles that are new to the map and we can only navigate about a meter of distance at a time.

**Reply by OfficialFlo** [reply]
@timothynassar We are currently also working with the MuSHR, and our autonomous navigation is also not working. Would it be possitble that you share your revision that is working? Would be great :)

**Reply by schmittlema** [reply]
There is a alpha version of mushr_rhc (just a different branch) intended to target py3 that we have been working on. To run, launch `real.launch` or `sim.launch`. That being said it is not intended to target the current version of the stack so there may be some issues. We are close to a noetic release including updated navigation code by end of June which should be more stable and reliable.

[mushr_rhc](https://github.com/prl-mushr/mushr_rhc/tree/noetic_rhc_replace)
